base/runtime_internals.jl

```diff
@@ -173,6 +173,58 @@ false
 """
 ispublic(m::Module, s::Symbol) = ccall(:jl_module_public_p, Cint, (Any, Any), m, s) != 0
 
+_function_macro_error() = (@noinline; error("@__FUNCTION__ can only be used within a function"))
+
+"""
+    @__FUNCTION__ -> Function
+
+Get the innermost enclosing function object.
+
+!!! note
+    In functions like `f() = [@__FUNCTION__ for _ in 1:10]`, this would
+    refer to the generator function in the comprehension, NOT the enclosing
+    function `f`. Similarly, in a closure function, this will refer to the
+    closure function object rather than the enclosing function.
```

Comment:
Maybe also worth mentioning that there's a lot of macros (`@spawn`, `@async`, etc.) that create closures as well.